src/metabase/api/api_key.clj

```diff
@@ -29,28 +30,83 @@
   (api/check-superuser)
   (api/checkp (not (t2/exists? :model/ApiKey :name name))
     "name" "An API key with this name already exists.")
-  (let [api-key (key-with-unique-prefix)
-        email   (format "api-key-user-%s@api-key.invalid" (u/slugify name))]
+  (let [unhashed-key (key-with-unique-prefix)
+        email        (format "api-key-user-%s@api-key.invalid" (u/slugify name))]
     (t2/with-transaction [_conn]
       (let [user (user/insert-new-user! {:email      email
                                          :first_name name
                                          :type       :api-key})]
         (user/set-permissions-groups! user [(perms-group/all-users) group_id])
-        (-> (t2/insert-returning-instance! :model/ApiKey
-                                           {:user_id      (u/the-id user)
-                                            :name         name
-                                            :unhashed_key api-key
-                                            :created_by   api/*current-user-id*})
-            (select-keys [:created_at :updated_at :id])
-            (assoc :name name
-                   :group_id group_id
-                   :unmasked_key api-key
-                   :masked_key (api-key/mask api-key)))))))
+        (let [api-key (-> (t2/insert-returning-instance! :model/ApiKey
+                                                         {:user_id      (u/the-id user)
+                                                          :name         name
+                                                          :unhashed_key unhashed-key
+                                                          :created_by   api/*current-user-id*})
+                          (t2/hydrate :group_name))]
+          (events/publish-event! :event/api-key-create
+                                 {:object  api-key
+                                  :user-id api/*current-user-id*})
+          (-> api-key
+              (select-keys [:created_at :updated_at :id :group_name :name])
+              (assoc :unmasked_key unhashed-key
+                     :masked_key (api-key/mask unhashed-key))))))))
 
 (api/defendpoint GET "/count"
   "Get the count of API keys in the DB"
   [:as _body]
   (api/check-superuser)
   (t2/count :model/ApiKey))
 
+(api/defendpoint PUT "/:id"
+  "Update an API key by changing either its group or its name"
+  [id :as {{:keys [group_id name] :as _body} :body}]
+  {id       ms/PositiveInt
+   group_id [:maybe ms/PositiveInt]
+   name     [:maybe ms/NonBlankString]}
+  (api/check-superuser)
+  (let [api-key-before (-> (t2/select-one :model/ApiKey :id id)
+                           ;; hydrate the group_name for audit logging
+                           (t2/hydrate :group_name))]
+    (when name
+      (t2/with-transaction [_conn]
+        ;; A bit of a pain to keep these in sync, but oh well.
+        (t2/update! :model/User (:user_id api-key-before) {:first_name name})
+        (t2/update! :model/ApiKey id {:name name})))
```

Comment:
I thought we were only going to store `name` on `ApiKey` and not on `User`? Or did that approach change after the product clarifications?

My bias would be to keep things normalized and only store it in one place, and it doesn't matter as much which place it is. But it's not a huge deal if you think this approach works OK.

Reply:
I'd prefer to store them in a single location, too. The pain point with that is:
- if we store them only in the ApiKey, we don't have a reliable way to populate the name. If someone does `(t2/select-one [:model/User :first_name :last_name :email] :id 1)`, it will have a `:common_name` that's incorrect, because they didn't also select the `:type` column.
- if we store them only in the User, we don't have a database-level invariant that they're unique. Not the end of the world, but it is nice to have.
To me, the downsides of either option outweigh the benefits of not having to keep things synced, but I'm not super tied to this approach.